File: src/main/java/kr/ac/kaist/jsaf/nodes_util/Span.java
/*******************************************************************************
    Copyright 2008,2010, Oracle and/or its affiliates.
    All rights reserved.


    Use is subject to license terms.

    This distribution may include materials developed by third parties.
 ******************************************************************************/

package kr.ac.kaist.jsaf.nodes_util;

import kr.ac.kaist.jsaf.useful.HasAt;
import kr.ac.kaist.jsaf.useful.MagicNumbers;
import kr.ac.kaist.jsaf.useful.NI;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;

public class Span implements Serializable, HasAt {
    public SourceLoc begin;
    public SourceLoc end;
    public CharVector cvec;

    @Override
    public int hashCode() {
        return begin.hashCode() * MagicNumbers.p + end.hashCode()
                * MagicNumbers.a;
    }

    @Override
    public boolean equals(Object o) {
        if (o instanceof Span) {
            Span sp = (Span) o;
            return begin.equals(sp.begin) && end.equals(sp.end);
        }
        return false;
    }

    public Span(SourceLoc b, SourceLoc e) {
        begin = b;
        end = e;
        cvec = new CharVector();
    }

    public Span addLines(int line) {
        return new Span(begin.addLines(line), end.addLines(line));
    }

    public static boolean beginsEarlierThan(Span a, Span b) {
        return (a.getBegin().getLine() < b.getBegin().getLine() ||
               (a.getBegin().getLine() == b.getBegin().getLine() &&
                a.getBegin().column() < b.getBegin().column()));
    }

    public static boolean endsLaterThan(Span a, Span b) {
        return (a.getEnd().getLine() > b.getEnd().getLine() ||
               (a.getEnd().getLine() == b.getEnd().getLine() &&
                a.getEnd().column() > b.getEnd().column()));
    }

    /** Span which includes both the given spans.  Assumption: they're
     * from the same file.  If this is not true, the results will be
     * unpredictable. */
    public Span(Span a, Span b) {
        if ( beginsEarlierThan(a,b) ) {
            begin = a.getBegin();
        } else {
            begin = b.getBegin();
        }
        if ( endsLaterThan(a,b) ) {
            end = a.getEnd();
        } else {
            end = b.getEnd();
        }
        cvec = a.getCharVector();
        cvec.merge(b.getCharVector().getVector());
    }

    /**
     * @return Returns the fileName. (contains full path "/c/safe_private/test.js")
     */
    public String getFileName() {
        return begin.getFileName();
    }

    /**
     * @return Returns the fileName. (contains only filename part "test.js")
     */
    public String getFileNameOnly() {
        return begin.getFileNameOnly();
    }

    public String convertNameSeparatorToSlash(String fileName) {
    	if(File.separatorChar == '/') return fileName;
    	else return fileName.replace(File.separatorChar, '/');
    }

    /**
     * @return Returns the begin.
     */
    public SourceLoc getBegin() {
        return begin;
    }

    /**
     * @return Returns the end.
     */
    public SourceLoc getEnd() {
        return end;
    }

    public CharVector getCharVector() {
        return cvec;
    }

    @Override
    public String toString() {
        try {
            return appendTo(new StringBuilder(), true).toString();
        } catch (IOException ex) {
            return NI.np();
        }
    }

    public String toStringWithoutFiles() {
        try {
            return appendTo(new StringBuilder(), false).toString();
        } catch (IOException ex) {
            return NI.np();
        }
    }

    public String at() {
        return toString();
    }

    public String stringName() {
        return "";
    }

    public Appendable appendTo(Appendable w, boolean do_files)
            throws IOException {
        return appendTo(w,do_files,false);
    }

    public Appendable appendTo(Appendable w, boolean do_files, boolean printer)
            throws IOException {
        int left_col = begin.column();
        int right_col = end.column();
        boolean file_names_differ =
            !(begin.getFileName().equals(end.getFileName()));
        do_files |= file_names_differ;

        if (printer) w.append(" @");
        if (do_files) {
            if (printer) w.append("\"");
            // Need to add escapes to the file name
            String beginFileName;
            if (printer) beginFileName = convertNameSeparatorToSlash(Unprinter.enQuote(begin.getFileName()));
            else beginFileName = convertNameSeparatorToSlash(begin.getFileName());
            w.append(beginFileName);
            if (printer) w.append("\"");
            w.append(":");
        }
        w.append(String.valueOf(begin.getLine()));
        w.append(":");
        w.append(String.valueOf(left_col));
        if (file_names_differ || begin.getLine() != end.getLine()
                || left_col != right_col) {
            w.append(printer ? Printer.tilde : "-");
            if (file_names_differ) {
                if (printer) w.append("\"");
                // Need to add escapes to the file name
                String endFileName;
                if (printer) endFileName = convertNameSeparatorToSlash(Unprinter.enQuote(end.getFileName()));
                else endFileName = convertNameSeparatorToSlash(end.getFileName());
                w.append(endFileName);
                if (printer) w.append("\"");
                w.append(":");
            }
            if (file_names_differ || begin.getLine() != end.getLine()) {
                w.append(String.valueOf(end.getLine()));
                w.append(":");
            }
            w.append(String.valueOf(right_col));
        }
        return w;
    }
}

File: src/main/scala/kr/ac/kaist/safe/safe_util/Span.scala
/**
 * *****************************************************************************
 * Copyright (c) 2016, KAIST.
 * All rights reserved.
 *
 * Use is subject to license terms.
 *
 * This distribution may include materials developed by third parties.
 * ****************************************************************************
 */

package kr.ac.kaist.safe.safe_util

import kr.ac.kaist.safe.useful.MagicNumbers
import kr.ac.kaist.safe.exceptions.SAFEError
import java.io.File
import java.io.IOException
import java.io.Serializable
import java.lang.StringBuilder

class Span(b: SourceLoc, e: SourceLoc) extends UIDObject with Serializable {
  var begin: SourceLoc = b
  var end: SourceLoc = e

  override def hashCode: Int =
    begin.hashCode * MagicNumbers.p + end.hashCode * MagicNumbers.a

  override def equals(o: Any): Boolean =
    if (o.isInstanceOf[Span]) {
      val sp: Span = o.asInstanceOf[Span]
      begin.equals(sp.begin) && end.equals(sp.end)
    } else false

  def addLines(line: Int, offset: Int): Span =
    new Span(begin.addLines(line, offset), end.addLines(line, offset))

  def beginsEarlierThan(a: Span, b: Span): Boolean =
    (a.begin.line < b.begin.line ||
      (a.begin.line == b.begin.line &&
        a.begin.column < b.begin.column))

  def endsLaterThan(a: Span, b: Span): Boolean =
    (a.end.line > b.end.line ||
      (a.end.line == b.end.line &&
        a.end.column > b.end.column))

  /**
   * Span which includes both the given spans.  Assumption: they're
   * from the same file.  If this is not true, the results will be
   * unpredictable.
   */
  def span(a: Span, b: Span): Unit = {
    if (beginsEarlierThan(a, b))
      begin = a.begin
    else
      begin = b.begin
    if (endsLaterThan(a, b))
      end = a.end
    else
      end = b.end
  }

  def fileName: String = begin.fileName

  def fileNameOnly: String = begin.fileNameOnly

  def convertNameSeparatorToSlash(fileName: String): String =
    if (File.separatorChar == '/') fileName
    else fileName.replace(File.separatorChar, '/')

  override def toString: String =
    try {
      appendToStr(new StringBuilder, true).toString
    } catch {
      case e: IOException => SAFEError.np.asInstanceOf[String]
    }

  def toStringWithoutFiles: String =
    try {
      appendToStr(new StringBuilder, false).toString
    } catch {
      case e: IOException => SAFEError.np.asInstanceOf[String]
    }

  override def at: String = toString

  def stringName: String = ""

  def appendToStr(w: StringBuilder, do_files: Boolean): StringBuilder = {
    val left_col = begin.column
    val right_col = end.column
    val file_names_differ = !(begin.fileName.equals(end.fileName))
    if (do_files | file_names_differ) {
      // Need to add escapes to the file name
      var beginFileName: String = ""
      beginFileName = convertNameSeparatorToSlash(begin.fileName)
      w.append(beginFileName)
      w.append(":")
    }
    w.append(String.valueOf(begin.line))
    w.append(":")
    w.append(String.valueOf(left_col))
    if (file_names_differ || begin.line != end.line || left_col != right_col) {
      w.append("-")
      if (file_names_differ) {
        // Need to add escapes to the file name
        var endFileName: String = ""
        endFileName = convertNameSeparatorToSlash(end.fileName)
        w.append(endFileName)
        w.append(":")
      }
      if (file_names_differ || begin.line != end.line) {
        w.append(String.valueOf(end.line))
        w.append(":")
      }
      w.append(String.valueOf(right_col))
    }
    w
  }
}
